src/peft/tuners/boft/layer.py

```diff
@@ -0,0 +1,578 @@
+        # Ensure the input matrix is skew-symmetric
+        skew = 0.5 * (data - data.transpose(1, 2))
+        I = torch.eye(r, device=data.device).unsqueeze(0).expand(b, r, c)
+
+        # Perform the Cayley parametrization
+        Q = torch.bmm(I - skew, torch.inverse(I + skew))
+        # Q = torch.linalg.solve(I + skew, I - skew, left=False)
+
+        return Q
+
+    def angle2rot(self, alphas):
+        """
+        Convert the angle to rotation matrix.
+        Only applicable for BOFT block size 2.
+        """
+        c = torch.cos(alphas)
+        s = torch.sin(alphas)
+        rot_mats = torch.cat([c, -s, s, c], dim=-1).view(alphas.shape[0], alphas.shape[1], 2, 2)
+        return rot_mats
+
+    def is_orthogonal(self, R, eps=1e-3):
+        """
+        Check if the matrix is orthogonal.
+        """
+        R = R.float()
+        with torch.no_grad():
+            RtR = torch.matmul(R.t(), R)
+            diff = torch.abs(RtR - torch.eye(R.shape[1], dtype=R.dtype, device=R.device))
+            return torch.all(diff < eps)
+
+    def is_identity_matrix(self, tensor):
+        """
+        Check if the matrix is identity.
+        """
+        if not torch.is_tensor(tensor):
+            raise TypeError("Input must be a PyTorch tensor.")
+        if tensor.ndim != 2 or tensor.shape[0] != tensor.shape[1]:
+            return False
+        identity = torch.eye(tensor.shape[0], device=tensor.device)
+        return torch.all(torch.eq(tensor, identity))
+
+    def forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            result_initilized = False  # Initialize the flag
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self.boft_R.keys():
+                    continue
+                boft_R = self.boft_R[active_adapter]
+                boft_s = self.boft_s[active_adapter]
+                dropout = self.boft_dropout[active_adapter]
+
+                N, D, H, _ = boft_R.shape
+                boft_R = boft_R.view(N * D, H, H)
+                orth_rotate_butterfly = self.cayley_batch(boft_R)
+                orth_rotate_butterfly = orth_rotate_butterfly.view(N, D, H, H)
+                orth_rotate_butterfly = dropout(orth_rotate_butterfly)
+                block_diagonal_butterfly = FastBlockDiag.apply(orth_rotate_butterfly)
+
+                butterfly_oft_mat_batch = torch.bmm(block_diagonal_butterfly, self.boft_P.permute(0, 2, 1))
+                butterfly_oft_mat_batch = torch.bmm(self.boft_P, butterfly_oft_mat_batch)
+                butterfly_oft_mat = butterfly_oft_mat_batch[0]
+
+                for i in range(1, butterfly_oft_mat_batch.shape[0]):
```

Comment:
@Zeju1997

Reply:
Yes, it is needed. Because the tests have quite toy network architecture (layer dimension quite small), it does not make sense to use the butterfly factorization to merge blocks. But in practise, this for loop will be called (if the number of butterfly factors is set to be bigger than 1).

Reply:
Added two tests in test_custom_models.py. Passed.

Reply:
This line is still not covered. When we look at the [line coverage](https://github.com/huggingface/peft/actions/runs/8416712762/job/23049579525?pr=1326), we get:

> src/peft/tuners/boft/layer.py                           465     76    84%   45, 57, 94-96, 100-102, 132, 196, 202-206, 209-216, 219-223, 234, 247, 253, 257, 265, 271, 275, 282, 290, 293, 319, 342, 377, 450, 489, 511-512, 541, 552, 571, 582, 593, 613-614, 644, 659, 665, 676, 682, 687-692, 700, 705-710, 717, 721-725, 751, 817-818, 857, 868, 889, 900, **911**, 942-943

As you can see, line 911 is not covered. I also checked locally with the debugger that this line is never run. Could you please check again that the tests you added really cover this line?

Reply:
Yes, you are right, the previous two tests were for the Linear layer. Additional two tests added to cover the Conv2d layer. Passed.

Reply:
Hmm, the latest test coverage result still shows this line as uncovered (e.g. [here](https://github.com/huggingface/peft/actions/runs/8423343811/job/23065084767?pr=1326#step:5:217)). But I'm fine with the testing overall, up to you if you want to check this further or not.